zkevm-circuits/src/evm_circuit/execution/error_oog_static_memory.rs

```diff
@@ -92,45 +167,155 @@ impl<F: Field> ExecutionGadget<F> for ErrorOOGStaticMemoryGadget<F> {
-        let is_mstore8 = self.is_mstore8.assign(
+        self.opcode
+            .assign(region, offset, Value::known(F::from(opcode.as_u64())))?;
+        self.is_mload.assign(
+            region,
+            offset,
+            F::from(opcode.as_u64()),
+            F::from(OpcodeId::MLOAD.as_u64()),
+        )?;
+        self.is_mstore8.assign(
             region,
             offset,
             F::from(opcode.as_u64()),
             F::from(OpcodeId::MSTORE8.as_u64()),
         )?;
 
         // Address in range check
-        self.address_in_range.assign(
+        let address_high = address_high::value::<F>(address.to_le_bytes());
+        self.address_in_range_high
+            .assign(region, offset, address_high)?;
+
+        let address_value = address_low::value(address.to_le_bytes());
+        let size = if opcode == OpcodeId::MSTORE8 { 1 } else { 32 };
+
+        let expanded_address = address_value
+            .checked_add(size)
+            .expect("address overflow u64");
+        self.expanded_address_in_range.assign(
             region,
             offset,
-            address_high::value::<F>(address.to_le_bytes()),
+            F::from(expanded_address),
+            F::from(1u64 << (N_BYTES_MEMORY_ADDRESS * 8)),
         )?;
 
-        // Memory expansion
-        self.memory_expansion.assign(
+        // TODO: sanity check, remove this after fixing #347 missing ErrGasUintOverflow
+        if address_high != F::zero() || expanded_address > (1u64 << (N_BYTES_MEMORY_ADDRESS * 8)) {
+            panic!("address overflow {} bytes", N_BYTES_MEMORY_ADDRESS);
+        }
+
+        let (_, memory_cost) = self.memory_expansion.assign(
             region,
             offset,
             step.memory_word_size(),
-            [address_low::value(address.to_le_bytes())
-                + if is_mstore8 == F::one() { 1 } else { 32 }],
+            [expanded_address],
         )?;
 
         // Gas insufficient check
         // Get `gas_available` variable here once it's available
-        self.insufficient_gas
-            .assign(region, offset, F::from(step.gas_cost - step.gas_left))?;
+        self.insufficient_gas.assign(
+            region,
+            offset,
+            F::from(step.gas_left),
+            F::from(OpcodeId::MLOAD.constant_gas_cost().as_u64() + memory_cost),
+        )?;
+
+        self.rw_counter_end_of_reversion.assign(
+            region,
+            offset,
+            Value::known(F::from(call.rw_counter_end_of_reversion as u64)),
+        )?;
+        self.restore_context
+            .assign(region, offset, block, call, step, 3)?;
 
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::test_util::CircuitTestBuilder;
+    use eth_types::{bytecode, word, ToWord};
+    use mock::test_ctx::helpers::account_0_code_account_1_no_code;
+    use mock::{eth, TestContext, MOCK_ACCOUNTS};
+
+    #[test]
+    fn test() {
+        let code = bytecode! {
+            PUSH32(0xFF)
```

Comment:
it is ok for now, if make pr against upstream, better to add `mstore8` test since it have different size .